Here you will find the supplementary material related to the ViRAE scientific paper.

**Supplementary figure 1:** Flowchart of the comparative virome methodology analysis performed in this study.

**Supplementary figure 2:** Histogram of the distribution of RNA-seq samples, which exhibited rRNA-virus chimeric phenomena in this study, based on the start and end positions of rRNA mapping, categorized according to the length percentile of rRNA-virus chimeric reads.

**Supplementary file 1:** Selection and filtering of single-end Next Generation Sequencing data related to viral metagenomics for the purposes of this study.

**Supplementary file 2:** Selection and filtering of paired-end Next Generation Sequencing data related to viral metagenomics for the purposes of this study.

**Supplementary file 3:** Metadata of the viruses analysed in this study.

**Supplementary file 4:** Analysis of ribosomal RNA contamination and comparative analysis of _de novo_ assembly efficacy performed by MEGAHIT software based on the metrics "UG50%" and "Max query covbases %", as calculated after the implementation of the 4 examined methods (RAW, BWA, SORTMERNA and ViRAE) separately on the studied RNA-seq dataset.

**Supplementary file 5:** Metadata of the viral sequences deposited by _Gil P. et al._ in the NCBI GenBank database along with their corresponding SRS data accession numbers in the NCBI SRA database.

**Supplementary file 6:** Enhanced viral genome _de novo_ assembly observed after the implementation of ViRAE on a small RNA-seq virome dataset published by _Gil P. et al_.

**Supplementary table 1:** Virus taxonomy and representation by RNA-seq samples in the analyzed data of this study.

**Supplementary table 2:** Frequency and taxonomy of the most contaminant species in the examined RNA-seq samples.
