Here you will find the supplementary material related to the ViRAE scientific paper.

**Supplementary figure 1:** Flowchart of the comparative virome methodology analysis performed in this study.

**Supplementary figure 2:** Flowchart of the pilot analysis related to the improvement of viral metagenomics papers.

**Supplementary figure 3:** Raincloud plot of the metric "True viral genome coverage (%)" computed via BLASTn for the proposed metagenomes associated with selected viral metagenomic research articles (noted as "NCBI") versus the MEGAHIT _de novo_ assembled sequences after ViRAE treatment of the corresponding SRS datasets from the same publications (noted as "ViRAE"). Each plot area consists of: i) half-violin shapes to show data distribution, colored according to the examined methods, and ii) a box plot for summary statistics. For better comprehension of the comparative analysis, light grey outline of the plotted half-violin shapes, corresponding to the performance of "ViRAE" method only, has been overlaid on top of the "NCBI" method. The mean and median values are also displayed within each box plot as “X” and “__” symbols respectively. The statistical analysis of the mean differences across the studied methods was performed by paired t-test with bonferroni adjustemnts, and the output statistical significance is indicated by (**** p-value < 0.0001).

**Supplementary file 1:** Selection and filtering of single-end Next Generation Sequencing data related to virome reasearch for the purposes of this study.

**Supplementary file 2:** Selection and filtering of paired-end Next Generation Sequencing data related to virome reasearch for the purposes of this study.

**Supplementary file 3:** Metadata of the viruses analysed in this study.

**Supplementary file 4:** Analysis of rRNA contamination and comparative analysis of _de novo_ assembly efficacy performed by MEGAHIT software based on specific metrics, as calculated after the implementation of the 4 examined methods (RAW, BWA, SORTMERNA and ViRAE) separately on the studied RNA-seq dataset.

**Supplementary file 5:** Metadata of the 5 randomly selected papers related to viral metagenomics research.

**Supplementary file 6:** Analysis of _de novo_ assembly improvement after the implementation of ViRAE on a pilot small-sized RNA-seq dataset sourced from recent publications related to viral metagenomics research.

**Supplementary table 1:** Virus taxonomy and representation by RNA-seq samples in the analyzed data of this study.

**Supplementary table 2:** Frequency and taxonomy of the most contaminant species in the examined RNA-seq samples.
