Here you will find the supplementary material related to the ZWA2 scientific paper.

**Supplementary figure 1:** Flowchart of the comparative virome methodology analysis performed in this study.

**Supplementary figure 2:** Comparison of the total number of reads mapped on viral sequences, as calculated after the implementation of each method (RAW, BWA, SORTMERNA and ZWA2) separately on the examined RNA-seq samples and subsequent mapping of the treated RNA-seq data on the corresponding viral genes/genomes by BWA. The median values for this metric are displayed within the bar plots. The statistical analysis of the mean differences across the studied methods was performed by RM-ANOVA and Tukey’s multiple comparison testing, and the output statistical significance of ZWA2 is indicated by (*** p-value < 0.001).

**Supplementary figure 3:** Comparison of the coverage of viral sequences, as calculated after the implementation of each method (RAW, BWA, SORTMERNA and ZWA2) separately on the examined RNA-seq samples and subsequent mapping of the treated RNA-seq data on the corresponding viral genes/genomes by BWA. The dotted lines connecting the colorful dots on the graph depict the fluctuations of the viral coverage rates for the same RNA-seq sample, treated separately by each method. A small subsection of the diagram (viral coverage rates between 95% and 100%) is zoomed in for representation purposes, clearly displaying the corresponding plotted data on the right side. The statistical analysis of the mean differences across the studied methods was performed by RM-ANOVA and Tukey’s multiple comparison testing, and the output statistical significance of ZWA2 is indicated by (** p-value < 0.01).

**Supplementary figure 4:** Comparison of the mean depth of reads mapped on viral sequences, as calculated after the implementation of each method (RAW, BWA, SORTMERNA and ZWA2) separately on the examined RNA-seq samples and subsequent mapping of the treated RNA-seq data on the corresponding viral genes/genomes by BWA. The median values for this metric are displayed within the bar plots. The statistical analysis of the mean differences across the studied methods was performed by RM-ANOVA and Tukey’s multiple comparison testing, and the output statistical significance of ZWA2 is indicated by (**** p-value < 0.0001).

**Supplementary file 1:** Selection and filtering of Next Generation Sequencing data related to viral metagenomics for the purposes of this study.

**Supplementary file 2:** Metadata of RNA-seq samples and their corresponding viruses analysed in this study.

**Supplementary file 3:** Analysis of ribosomal RNA contamination in the examined RNA-seq samples of this study.

**Supplementary file 4:** Comparative analysis of virus mapping performance based on the metrics "Virus mapped reads", "Virus coverage (%)", and "Mean depth of virus mapped reads", as calculated after the implementation of the 4 examined methods (RAW, BWA, SORTMERNA and ZWA2) separately on the studied RNA-seq samples and subsequent mapping of the treated RNA-seq data on the corresponding viral genes/genomes by BWA software.

**Supplementary file 5:** Comparative analysis of _de novo_ assembly performance based on the metrics "UG50%" and "Max query covbases %", as calculated after the implementation of the 4 examined methods (RAW, BWA, SORTMERNA and ZWA2) separately on the studied RNA-seq samples and subsequent _de novo_ assembly of the treated RNA-seq data by MEGAHIT software.

**Supplementary file 6:** Enhanced viral genome assembly observed after the implementation of ZWA2 on a small RNA-seq virome dataset.

**Supplementary Table 1:** Comparison of the pilot version of ZWA (ZWA1) versus the evolved ZWA2 version in terms of software features as well as virus mapping and de novo assembly performance based on the _Anopheles sacharovi_ mosquito RNA-seq sample (SRA accession number SRR13449040), its 28S and 18S rRNAs given as reference (GenBank accession numbers MT808434 & MT808462), and Xanthi rhabdovirus (GenBank
accession number MW520377) genome used for testing.

**Supplementary table 2:** Virus taxonomy and representation by RNA-seq samples in the analyzed data of this study.

**Supplementary table 3:** Frequency and taxonomy of the most contaminant species in the examined RNA-seq samples.
